### PR TITLE
Use block requests to check if block responses are correct

### PR DESCRIPTION
**Pull request by bkchr**
Before this pr sync relied on recently announced blocks to check if a
given peer response is correct. However this could lead to situations
where we requested a block from a peer and it gave us the requested, but
we rejected the response because this peer never send us an announcement
for the given block. See the added tests for a reproduction of the
problem.
With this pr, we now take the block request to check if a given response
matches the request. A node should not send us a block response
without a request anyway.

Essentially there is still a bug, because as you see in the test, we are
requesting block 2, while we already have this block imported. It even
happens that we request a block from the network that we have authored.
However a fix for this would require some more refactoring of the sync code.

### CODE REVIEW

File: client/network/src/block_requests.rs [hunk 1]
```diff
@@ -336,12 +336,11 @@ where
...
-		( &mut self
-		, peer: &PeerId
-		, request: &schema::v1::BlockRequest
-		) -> Result<schema::v1::BlockResponse, Error>
-	{
+	fn on_block_request(
+		&mut self,
+		peer: &PeerId,
+		request: &schema::v1::BlockRequest,
+	) -> Result<schema::v1::BlockResponse, Error> {
```
Comment (tomaka) [inline client/network/src/block_requests.rs:343]: We're removing this in https://github.com/paritytech/substrate/pull/7478
I'd suggest to revert these changes to not cause merge conflicts

File: client/network/src/protocol/sync.rs [hunk 2]
```diff
@@ -1846,4 +1862,168 @@ mod test {
...
+
+		let request = requests[0].1.clone();
+
+		assert_eq!(from, request.from);
+		assert_eq!(Some(max), request.max);
+		request
+	}
+
+	#[test]
+	fn lol() {
```
Comment (tomaka) [inline client/network/src/protocol/sync.rs:1929]: I'd suggest a more explicit test name

Comment (bkchr) [reply client/network/src/protocol/sync.rs:1929]: Ohh fuck :see_no_evil: Now I know what I have forgotten.

**Comment by arkpar** [comment]
> Essentially there is still a bug, because as you see in the test, we are
requesting block 2, while we already have this block imported.

In some cases this is inevitable. If a peer suddenly sends an announcement for a future block that we don't know anything about, we request a chain starting from the last known common block. If they did not announce that they now have intermediate blocks, we can't be sure if that future block is continuing a known chain. 

It looks like in your tests some peers skip some announcements. I. e. Peer1 was on block 1 but suddenly sends and announcement for block 4? During normal operation this should not happen. All blocks are announced to all peers, unless full sync mode.

File: client/network/src/protocol/sync.rs [hunk 3]
```diff
@@ -1685,20 +1668,60 @@ fn validate_blocks<Block: BlockT>(blocks: &Vec<message::BlockData<Block>>, who:
 			}
 		}
 	}
+
+	if let Some(request) = request {
```
Comment (arkpar) [inline client/network/src/protocol/sync.rs:1672]: This should be checked first, before doing more expensive checks, such as computing trie roots.
Also, some of the checks against request are being performed on the protocol level here: 
https://github.com/paritytech/substrate/blob/master/client/network/src/protocol.rs#L709
Might be a good idea to put them in one place.

**Comment by bkchr** [comment]
> It looks like in your tests some peers skip some announcements. I. e. Peer1 was on block 1 but suddenly sends and announcement for block 4? During normal operation this should not happen. All blocks are announced to all peers, unless full sync mode.

I can give you a trace. This is exactly what the logs showed :shrug:

**Comment by arkpar** [comment]
> I can give you a trace. This is exactly what the logs showed 🤷

Would like to check it. Might be a further problem with how announcements are sent out or ignored when received.  

Don't get me wrong, this is still a good test as we should account for all kinds of misbehaviour. And peers might legitimately skip announcements if they fall in to full sync mode for some reason, but this should not happen regularly. And if they do skip announcements and make us download extra blocks, we should not disconnect, but decrease reputation by a bit. This is what the current code does.

**Comment by bkchr** [comment]
Yeah :) The problem we have seen is that peers are getting disconnected really fast with latest polkadot master. I will send you the log

**Comment by arkpar** [comment]
Examining the logs, I guess the issue is that we don't receive back announcements for blocks that we have already announced to peers (e.g. blocks that are collated by this node). So the original issue is that recently_announced should really include not only received, but sent out announcements as well. But since you've removed it anyway, is should be OK

Looks good, modulo @tomaka's concerns.

**Review by tomaka — APPROVED** [review]
No objection from me

**Review by arkpar — APPROVED** [review]
